Path normalization fails in misago

The routing in https://github.com/rafalp/Misago confuses the current path normalization, causing it to fail:

```
path_info = '/a/secret/1/'
resolved = ResolverMatch(func=misago.threads.views.attachment.attachment_server, args=(), kwargs={'secret': 'secret', 'pk': '1'}, url_name=attachment, app_names=['misago'], namespaces=['misago'], route=^a/(?P<secret>[-a-zA-Z0-9]+)/(?P<pk>\d+)/)

    def normalize_path_info(path_info, resolved):
        if not resolved.kwargs:
            # No kwargs mean no parameters to normalize
            return path_info

        resolver = get_resolver()
>       regex, match = get_resolved_match(path_info, resolver)
E       TypeError: cannot unpack non-iterable NoneType object

/src/appmap-python/appmap/django.py:160: TypeError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   extensions.
 
 ### Fixed
+- [#138] Path normalization for django requests is now more robust.
 - [#128] The error message for a missing config file now shows the full path where the
   file was expected to be found.
 

--- a/appmap/django.py
+++ b/appmap/django.py
@@ -14,6 +14,8 @@ from django.dispatch import receiver
 from django.http import HttpResponse, HttpResponseBadRequest, JsonResponse
 from django.template import Template
 from django.urls import get_resolver, resolve
+from django.urls.resolvers import _route_to_regex
+from django.urls.exceptions import Resolver404
 
 import time
 import json
@@ -140,16 +142,31 @@ def request_params(request):
     return values_dict(params.lists())
 
 def get_resolved_match(path_info, resolver):
-    # Look through the resolver's cache of reverse
-    # route mappings and find the regex that matches the current
-    # request. The route mappings are tuples of the form:
-    #   (format string, regex, args, kwargs)
-    regex = None
-    for _,values in resolver.reverse_dict.lists():
-        for _,regex,_,_ in values:
-            match = re.match(regex, path_info[1:])
-            if match:
-                return (regex, match)
+    resolver_match = resolver.resolve(path_info)
+
+    # Sometimes the route returned by resolve is a regex, sometimes
+    # it's a url pattern.
+
+    # Start by assuming it's a regex.
+    regex = resolver_match.route
+
+    match = re.match(regex, path_info[1:])
+    if match:
+        return (regex, match)
+
+    # If it didn't match, it's a url pattern. Use _route_to_regex to
+    # turn it into a regex.  (url patterns sometimes start with a
+    # caret, which needs to be stripped before conversion.)
+    if regex[0] == '^':
+        regex = regex[1:]
+    regex = _route_to_regex(regex)[0]
+
+    match = re.match(regex, path_info[1:])
+    if not match:
+        raise RuntimeError('No match for %s found with resolver %r, regex %s' %
+                           (path_info, resolver, regex))
+
+    return (regex, match)
 
 def normalize_path_info(path_info, resolved):
     if not resolved.kwargs:
@@ -158,12 +175,6 @@ def normalize_path_info(path_info, resolved):
 
     resolver = get_resolver()
     regex, match = get_resolved_match(path_info, resolver)
-
-    # Given that we're processing a resolved request, there must be
-    # one that matches.
-    if not match:
-        raise RuntimeError('No match for %s found with resolver %r' %
-                           (path_info, resolver))
 
     # Compile the regex so we can see what groups it contains.
     groups = re.compile(regex).groupindex
@@ -195,14 +206,21 @@ class Middleware:
         if self.recorder.enabled:
             add_metadata()
             start = time.monotonic()
-            resolved = resolve(request.path_info)
             params = request_params(request)
-            params.update(resolved.kwargs)
+            try:
+                resolved = resolve(request.path_info)
+                params.update(resolved.kwargs)
+                normalized_path_info = normalize_path_info(request.path_info, resolved)
+            except Resolver404:
+                # If the request was for a bad path (e.g. when an app
+                # is testing 404 handling), resolving will fail.
+                normalized_path_info = None
+
             call_event = HttpServerRequestEvent(
                 request_method=request.method,
                 path_info=request.path_info,
                 message_parameters=params,
-                normalized_path_info=normalize_path_info(request.path_info, resolved),
+                normalized_path_info=normalized_path_info,
                 protocol=request.META['SERVER_PROTOCOL'],
                 headers=request.headers
             )

--- a/appmap/test/test_django.py
+++ b/appmap/test/test_django.py
@@ -1,6 +1,8 @@
 # flake8: noqa: E402
 # pylint: disable=unused-import, redefined-outer-name, missing-function-docstring
 
+from django.urls import get_resolver, resolve
+
 from pathlib import Path
 
 import django
@@ -12,6 +14,8 @@ from django.template.loader import render_to_string
 import django.test
 from django.test.client import MULTIPART_CONTENT
 from django.urls import include, path, re_path
+from django.conf.urls import url
+
 import pytest
 
 import appmap
@@ -122,6 +126,18 @@ def exception_view(_request):
 def user_included_view(_request, username):
     return django.http.HttpResponse(f'user {username}, included')
 
+def acl_edit(_request, pk):
+    return django.http.HttpResponse(f'acl_edit {pk}')
+
+# replicate a problematic bit of misago's routing
+admincp_patterns = [
+    url('^', include([
+        url(r'^permissions/', include(([
+            url(r'^edit/(?P<pk>\d+)$', acl_edit, name='edit')
+        ], "permissions"), namespace="permissions"))
+    ]))
+]
+
 urlpatterns = [
     path('test', view),
     path('', view),
@@ -132,9 +148,10 @@ urlpatterns = [
     path('echo', echo_view),
     path('exception', exception_view),
     re_path(r'^post/included/', include([
-            path('<username>', user_view)]))
-]
+            path('<username>', user_view)])),
 
+    url(r'^admincp/', include((admincp_patterns, 'admin'), namespace="admin"))
+]
 
 @pytest.mark.appmap_enabled
 def test_unnamed(client, events):
@@ -175,6 +192,15 @@ def test_exception(client, events, monkeypatch):
             'class': 'builtins.RuntimeError',
             'message': 'An error'
     })]
+
+@pytest.mark.appmap_enabled
+def test_deeply_nested_routes(client, events):
+    client.get('/admincp/permissions/edit/1')
+
+    assert len(events) == 2
+    assert events[0].http_server_request == DictIncluding({
+        'normalized_path_info': '/admincp/permissions/edit/{pk}'
+    })
 
 django.conf.settings.configure(
     DATABASES={'default': {'ENGINE': 'django.db.backends.sqlite3', 'NAME': ':memory:'}},

